07-generic_bookworms/final/books/book.go

```diff
@@ -0,0 +1,32 @@
+package books
+
+import (
+	"fmt"
+
+	"learngo-pockets/genericworms/collectors"
+)
+
+// Book describes a book on a collector's shelf.
+type Book struct {
+	Author string `json:"author"`
+	Title  string `json:"title"`
+}
+
+func (b Book) Before(sortable collectors.Sortable) bool {
```

Comment:
This doesn't look right to me, the method declaration here should be:

```
func (b Book) Before(other Book) bool {
```

You can achieve this with two approaches:

```
package main

import (
	"fmt"
	"sort"
)

type Sortable[T any] interface {
	Before(T) bool
}

func sortItems[T Sortable[T]](items []T) []T {
	sort.Slice(items, func(i, j int) bool {
		return items[i].Before(items[j])
	})
	return items
}

func sortItems2[T any](xs []T, less func(T, T) bool) {
	sort.Slice(xs, func(i, j int) bool {
		return less(xs[i], xs[j])
	})
}

type Book struct {
	Name string
}

func (b Book) Before(other Book) bool {
	return b.Name < other.Name
}

func main() {
	books1 := []Book{{"B"}, {"C"}, {"A"}}
	sortItems(books1)
	fmt.Println(books1)

	books2 := []Book{{"B"}, {"C"}, {"A"}}
	sortItems2(books2, Book.Before)
	fmt.Println(books2)
}
```

The `sortItems2` is somewhat nicer, because it allows easily to vary the sorting func. For example https://pkg.go.dev/golang.org/x/exp/slices#SortFunc uses the same approach.

Reply:
very minor nit: Also, name `Less` probably makes more sense than `Before`. Mainly because `Before` is used for time comparison, it could make seem like it would sort them by date of publication.